### ISSUE TITLE
Dummy variables containing an underscore (_) are *not* ignored.

### ISSUE THREAD

**Issue by sanscore** [PyCQA/pylint#1058]
### Steps to reproduce
1. Have a "dummy" variable that has an underscore in it's root name.
  * __Root Name:__ The variable name without the dummy prefix, underscore (`_`).
  * __Example:__ root: `unused_var`, dummy-fied: `_unused_var`. 
1. Run code against pylint, it will complain that `_unused_var` is an `unused_variable` (or an `unused-argument`).

*Note:* Discovered when investigating https://github.com/PyCQA/pylint/issues/1057

### Current behavior

Pylint will __not__ ignore "dummy" variables (a variable that has been prefixed with an underscore) when they have an underscore (`_`) elsewhere in their name.

#### Unused Argument

Pylint reports `[unused-argument] Unused argument '_never_used'` for:

```python
def test_unused_arg(_unused_arg):
    """_unused_arg should be ignored, but it isn't."""
    assert False
```

#### Unused Variable

Pylint reports `[unused-variable] Unused variable '_unused_var'` for:

```
def test_unused_var():
    """_unused_var should be ignored, but it isn't."""
    _unused_var = True
    assert False
```

### Expected behavior

All "dummy" variables should be ignored. The problems comes down the default regular express for `dummy-variables-rgx`.

__Current default setting:__
```
dummy-variables-rgx=(_+[a-zA-Z0-9]*?$)|dummy
```

__Proposed change:__
The proposed change, simply, adds an underscore.
```
dummy-variables-rgx=(_+[a-zA-Z0-9_]*?$)|dummy
```

### pylint --version output

```
pylint 1.6.4,
$ pylint --version
astroid 1.4.7
Python 2.7.11 (default, Mar 15 2016, 09:47:58)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

**Comment by The-Compiler** [comment]
Sounds good to me - want to open a pull request? The default is set in `VariablesChecker` in `pylint/checkers/variables.py`.

**Comment by sanscore** [comment]
Certainly. I've created the commit, but that one character change was enough to cause two failures.

### function_redefined.py failure

The first failure is a bit baffling. ... I'm without a clue as to why my change has regressed this test.

pylint/test/functional/function_redefined.py (3, 32-42):
```
__revision__ = ''
...
def func2(): # [function-redefined]
    """docstring"""
    __revision__ = 1 # [redefined-outer-name]
    return __revision__

if __revision__:
    def exclusive_func():
        "docstring"
else:
    def exclusive_func():
        "docstring"
```

pylint/test/functional/function_redefined.txt:
```
function-redefined:15:AAAA.method2:method already defined line 12
function-redefined:18:AAAA:class already defined line 5
function-redefined:32:func2:function already defined line 29
redefined-outer-name:34:func2:Redefining name '__revision__' from outer scope (line 3)
function-redefined:51:exclusive_func2:function already defined line 45
```

Is there some sort of order, or precedence, for checkers!? My assumption is that VariablesChecker permitted the `__revision__` token, so then the checker responsible for `[redefined-outer-name]` was skipped or ignored.

### unused_variable.py failure
The second failure, is expected.

pylint/test/functional/unused_variable.py (10-16):
```python
def test_unused_with_prepended_underscore():
    _foo = 42
    _ = 24
    __a = 24
    dummy = 24
    _a_ = 42 # [unused-variable]
    __a__ = 24 # [unused-variable]
```

pylint/test/functional/unused_variable.txt:
```
unused-variable:4:test_regression_737:Unused variable 'xml'
unused-variable:7:test_regression_923:Unused variable 'unittest.case'
unused-variable:8:test_regression_923:Unused variable 'sql'
unused-variable:15:test_unused_with_prepended_underscore:Unused variable '_a_'
unused-variable:16:test_unused_with_prepended_underscore:Unused variable '__a__'
```

The failures were consistent through all python versions (pypy and jython, not tested).

### tox output
```
======================================================================
FAIL: function_redefined (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py33/lib/python3.3/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "function_redefined":

Expected in testdata:
  34: redefined-outer-name

======================================================================
FAIL: unused_variable (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py33/lib/python3.3/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "unused_variable":

Expected in testdata:
  15: unused-variable
  16: unused-variable

----------------------------------------------------------------------
Ran 716 tests in 35.497s
```

**Comment by sanscore** [comment]
In regards to the `unused_variable.py` failure, assuming that it is necessary to keep those two failures, then I can just correct `dummy-variables-rgx`.

Given a setting that fixes the `unused_variable.py` failure, I would expect to also fix the `function_redfined.py` error as well. However, to me, this regression seems to indicate a peculiar coupling between the two checkers.

**Comment by The-Compiler** [comment]
Hmm, I feel like I can't really help here, sorry. Let's wait until @PCManticore wakes up :wink:

**Comment by sanscore** [comment]
Attempted this alternative: `(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy`

Fixes the problems experienced in https://github.com/PyCQA/pylint/issues/1058#issuecomment-237162725, but causes new issues. I'll investigate the new issues of, previously, expected `unused-variable` to see if they no longer meet those criteria. My thinking being, maybe these other tests tested -- in my opinion -- the existing bad behavior. Or, the more likely possibility, the new regexp has borked good tests. 

I'll report back.

#### tox output

```
======================================================================
FAIL: test_functionality (pylint.testutils.LintTC)
module test of input file "func_no_dummy_redefined" (LintTC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/testutils.py", line 273, in test_functionality
    self._test(tocheck)
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/testutils.py", line 293, in _test
    self._check_result(self.linter.reporter.finalize())
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/testutils.py", line 277, in _check_result
    got.strip()+'\n')
AssertionError: 'C:  7: Invalid constant name "value"\nW: 12:clobbering: Redefining name \'value [truncated]... != 'C:  7: Invalid constant name "value"\nW: 12:clobbering: Redefining name \'value [truncated]...
  C:  7: Invalid constant name "value"
  W: 12:clobbering: Redefining name 'value' from outer scope (line 7)
+ W: 13:clobbering: Redefining name '_' from outer scope (line 5)
+ W: 13:clobbering: Unused variable '_'


======================================================================
FAIL: lost_exception (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "lost_exception":

Unexpected in testdata:
  29: unused-variable

======================================================================
FAIL: not_in_loop (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "not_in_loop":

Unexpected in testdata:
  37: redefined-outer-name

======================================================================
FAIL: reused_outer_loop_variable (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "reused_outer_loop_variable":

Unexpected in testdata:
  36: redefined-outer-name

======================================================================
FAIL: too_many_statements (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "too_many_statements":

Unexpected in testdata:
  30: unused-variable

======================================================================
FAIL: unused_variable (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "unused_variable":

Unexpected in testdata:
  12: unused-variable

======================================================================
FAIL: useless_else_on_loop (test_functional.LintModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/u205/work/pylint/.tox/py27/lib/python2.7/site-packages/pylint/test/test_functional.py", line 314, in _runTest
    self.fail('\n'.join(msg))
AssertionError: Wrong results for file "useless_else_on_loop":

Unexpected in testdata:
  24: unused-variable
  49: unused-variable

----------------------------------------------------------------------
Ran 723 tests in 31.865s

FAILED (failures=7, skipped=58)
```

**Comment by sanscore** [comment]
My new regexp was borking on `_` which seems obvious in retrospect.

**Comment by sanscore** [comment]
Amending my previous regexp `_+$|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy` results in no errors. I'll contribute a couple lines to `unused_variable.py` and `unused_argument.py` to ensure the suggested behavior remains consistent.

**Comment by dwaynebailey** [comment]
If it helps our policy in our code is `__` or `$var_` for unused variables and we're using this regex.  The last part is likely a simpler regex to achieve what is needed here for `_$var`

`dummy-variables-rgx=__|^.*[^_]_$`
